### ISSUE TITLE
Event delegation does not match jQuery behaviour

### ISSUE THREAD

**Issue by KaptajnKold** [HubSpot/youmightnotneedjquery#316]
The proposed vanilla version of jQuery's `$(selector).on(eventName, selector, eventHandler)` does not match jQuery's behaviour:

```javascript
function addEventListener(el, eventName, eventHandler, selector) {
  if (selector) {
    const wrappedHandler = (e) => {
      if (e.target && e.target.matches(selector)) {
        eventHandler(e);
      }
    };
    el.addEventListener(eventName, wrappedHandler);
    return wrappedHandler;
  } else {
    el.addEventListener(eventName, eventHandler);
    return eventHandler;
  }
}
```

Consider the following HTML:
```html
<ul class=thumbnails>
  <li>
    <img src=whatever />
  </li>
  <li>
    <img src=whatever />
  </li>
  <li>
    <img src=whatever />
  </li>
</ul>
```

in jQuery, if I do the following:

```javascript
$(document).on('click', '.thumbnails li', function () {
  $(this).toggleClass('selected');
});
```

... I can expect that clicking on an image will toggle the `selected` class on the list item that contains the image.

If I try to do the same with the proposed solution, nothing happens. This is because the event target is the image element, and not the list item element.

To make a drop-in replacement for jQuery, the code below will work:

```javascript
function addEventListener(el, eventName, eventHandler, selector) {
  if (selector) {
    const wrappedHandler = (e) => {
      if (!e.target) return)
      const el = e.target.closest(selector)
      if (el) {
        eventHandler.call(el, e);
      }
    };
    el.addEventListener(eventName, wrappedHandler);
    return wrappedHandler;
  } else {
    const wrappedHandler = (e) => {
      eventHandler.call(el, e);
    }
    el.addEventListener(eventName, wrappedHandler);
    return wrappedHandler;
  }
}

```

**Comment by Abbondanzo** [comment]
Good catch! I made a slight adjustment as part of #330 to your provided fix that overwrites the event's target should the user not want to use a function scoped to the element, but otherwise encompasses this change

**Comment by nk-o** [comment]
Hi @Abbondanzo. It is not working when trying to call `e.preventDefault();`.

I see this error in Chrome: `Uncaught TypeError: Illegal invocation`

![Shot 2023-07-03 at 11 25 31](https://github.com/HubSpot/youmightnotneedjquery/assets/14810891/17b015c5-e4b7-40c3-9cf2-39d62d311640)

**Comment by Abbondanzo** [comment]
> Hi @Abbondanzo. It is not working when trying to call `e.preventDefault();`.
> 
> I see this error in Chrome: `Uncaught TypeError: Illegal invocation`
> 
> ![Shot 2023-07-03 at 11 25 31](https://user-images.githubusercontent.com/14810891/250542734-17b015c5-e4b7-40c3-9cf2-39d62d311640.png)

Good call out! I've patched this away (mostly) in #346. There's a slight shortcoming in that the `currentTarget` on vanilla JS will always be the element you've attached your listener to, and I want to find a simple solution that allows mutation of that property, but the patch provides a semi-working solution in the interim
